[SPARK-6485] [MLlib] [Python] Add CoordinateMatrix/RowMatrix/IndexedRowMatrix to PySpark.

dusenberrymw opened the pull request:
This PR adds the RowMatrix, IndexedRowMatrix, and CoordinateMatrix distributed matrices to PySpark.  Each distributed matrix class acts as a wrapper around the Scala/Java counterpart by maintaining a reference to the Java object.  New distributed matrices can be created using factory methods added to DistributedMatrices, which creates the Java distributed matrix and then wraps it with the corresponding PySpark class.  This design allows for simple conversion between the various distributed matrices, and lets us re-use the Scala code.  Serialization between Python and Java is implemented using DataFrames as needed for IndexedRowMatrix and CoordinateMatrix for simplicity.  Associated documentation and unit-tests have also been added.  To facilitate code review, this PR implements access to the rows/entries as RDDs, the number of rows & columns, and conversions between the various distributed matrices (not including BlockMatrix), and does not implement the other linear algebra functions of the matrices, although this will be very simple to add now.

dusenberrymw committed:
Implemented the RowMatrix API in PySpark by doing the following: Added a DistributedMatrices class to contain factory methods for creating the various distributed matrices.  Added a factory method for creating a RowMatrix from an RDD of Vectors.  Added a createRowMatrix function to the PythonMLlibAPI to interface with the factory method.  Added DistributedMatrix, DistributedMatrices, and RowMatrix classes to the pyspark.mllib.linalg api.

dusenberrymw committed:
Adding unit tests for RowMatrix methods.

dusenberrymw committed:
Updating design to have a PySpark RowMatrix simply create and keep a reference to a wrapper over a Java RowMatrix.  Updating DistributedMatrices factory methods to accept numRows and numCols with default values.  Updating PySpark DistributedMatrices factory method to simply create a PySpark RowMatrix. Adding additional doctests for numRows and numCols parameters.

dusenberrymw committed:
Updating documentation to add PySpark RowMatrix. Inserting newline above doctest so that it renders properly in API docs.

dusenberrymw committed:
Implemented the IndexedRowMatrix API in PySpark, following the idea of the RowMatrix API.  Note that for the IndexedRowMatrix, we use DataFrames to serialize the data between Python and Scala/Java, so we accept PySpark RDDs, then convert to a DataFrame, then convert back to RDDs on the Scala/Java side before constructing the IndexedRowMatrix.

dusenberrymw committed:
Updating the architecture a bit to make conversions between the various distributed matrix types easier.  The different distributed matrix classes are now only wrappers around the Java objects, and take the Java object as an argument during construction.  This way, we can call  for example on an , which returns a reference to a Java RowMatrix object, and then construct a PySpark RowMatrix object wrapped around the Java object.  This is analogous to the behavior of PySpark RDDs and DataFrames.  We now delegate creation of the various distributed matrices from scratch in PySpark to the factory methods on .

dusenberrymw committed:
Implemented the CoordinateMatrix API in PySpark, following the idea of the IndexedRowMatrix API, including using DataFrames for serialization.

dusenberrymw committed:
Added wrappers for the conversions between the various distributed matrices.  Added logic to be able to access the rows/entries of the distributed matrices, which requires serialization through DataFrames for IndexedRowMatrix and CoordinateMatrix types. Added unit tests.

dusenberrymw committed:
Updating unit test to be more useful.

dusenberrymw committed:
Updating documentation for each of the distributed matrices.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37893 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37893/consoleFull) for   PR 7554 at commit [`06c39e2`](https://github.com/apache/spark/commit/06c39e215028584fc2b3849ee5be57fc2ee2a373).

SparkQA commented:
[Test build #37893 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37893/console) for   PR 7554 at commit [`06c39e2`](https://github.com/apache/spark/commit/06c39e215028584fc2b3849ee5be57fc2ee2a373).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class DistributedMatrices(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(object):`
  * `trait ExpectsInputTypes extends Expression `
  * `trait ImplicitCastInputTypes extends ExpectsInputTypes `
  * `trait Unevaluable extends Expression `
  * `trait Nondeterministic extends Expression `
  * `trait CodegenFallback extends Expression `
  * `case class Ascii(child: Expression) extends UnaryExpression with ImplicitCastInputTypes `
  * `case class Base64(child: Expression) extends UnaryExpression with ImplicitCastInputTypes `
  * `case class UnBase64(child: Expression) extends UnaryExpression with ImplicitCastInputTypes `
  * `  case class FakeFileStatus(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

dusenberrymw committed:
Fixing long lines.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37898 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37898/consoleFull) for   PR 7554 at commit [`1b8ea74`](https://github.com/apache/spark/commit/1b8ea7473999d90580a2524dff584482dd56108c).

SparkQA commented:
[Test build #37898 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37898/console) for   PR 7554 at commit [`1b8ea74`](https://github.com/apache/spark/commit/1b8ea7473999d90580a2524dff584482dd56108c).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class DistributedMatrices(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(object):`
  * `trait ExpectsInputTypes extends Expression `
  * `trait ImplicitCastInputTypes extends ExpectsInputTypes `
  * `trait Unevaluable extends Expression `
  * `trait Nondeterministic extends Expression `
  * `trait CodegenFallback extends Expression `
  * `  case class FakeFileStatus(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

MechCoder commented:
I'll try to give a first pass later today.

mengxr commented:
@dusenberrymw Thanks for working on this feature! Please fix the following:

1. Do not introduce methods in Python that are not available on the Scala side, e.g. `DistributeMatrices`. If we want to add new methods, we should add them to Scala first and then Python wrappers. If those methods are useful, we can do that in a follow-up PR.
2. Check code style. For Scala code, we follow https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide. For Python, we follow PEP8. I saw many long lines in your Python code. Having a more consistent code style would make code review easier.

dusenberrymw committed:
Reformatting a few long Python doctest lines.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37981 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37981/consoleFull) for   PR 7554 at commit [`88f28c4`](https://github.com/apache/spark/commit/88f28c44046b9c3f57cbaf3b0daf62e14bd05ee1).

SparkQA commented:
[Test build #37981 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37981/console) for   PR 7554 at commit [`88f28c4`](https://github.com/apache/spark/commit/88f28c44046b9c3f57cbaf3b0daf62e14bd05ee1).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class DistributedMatrices(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(object):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

dusenberrymw commented:
@mengxr No problem, it has been enjoyable to work on!  Here are some thoughts:

1. For local matrices and vectors, Scala has `Matrices` and `Vectors` classes, which each contain factory methods for creating the various local `Matrix` and `Vector` types (`DenseVector`, `SparseVector`, `DenseMatrix`, `SparseMatrix`).  These factory methods are the recommended method for creating these matrices & vectors.  On the Python side, there are also `Matrices` and `Vectors` classes with factory methods, however, rather than call the Scala counterpart, these just mimic the behavior and create the various `Matrix` and `Vector` types directly in Python.  For the _distributed_ matrices, I thought it would be best to follow the same idea, so I added a `DistributedMatrices` class in Scala containing factory methods, and created the equivalent in Python.  On the Python side, I think this ends up being a really clean solution, as it allows the specific types of distributed matrices (`RowMatrix`, `IndexedRowMatrix`, etc.) to simply be wrappers over their Scala/Java counterpart, similar to how the RDD and DataFrame classes act in Python.  This keeps the creation logic within the factory methods, and allows for clean conversions between the distributed matrix types in Python.  Really interested in your thoughts on this!  I'd definitely be willing to pull that out into a separate pull request though, should that end up being the best idea.
2. For point number _2_, yes, there were still a few long Python doctests, but I have cleaned those up now!  Also, looks like I'm now having issues with the unit tests between Python2 & Python3 (`2` vs `2L` for example), so I need to look into that.  The logic is correct, but just need to fix the output so that both Python versions output the same.

dusenberrymw committed:
Updating Python doctests for compatability between Python 2 & 3. Since Python 3 removed the idea of a separate 'long' type, all values that would have been outputted as a 'long' (ex: '4L') will now be treated as an 'int' and outputed as one (ex: '4').  The doctests now explicitly convert to ints so that both Python 2 and 3 will have the same output.  This is fine since the values are all small, and thus can be easily represented as ints.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37991 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37991/consoleFull) for   PR 7554 at commit [`f0c19a0`](https://github.com/apache/spark/commit/f0c19a015778e5086b50398d7e24787f80b68c0d).

SparkQA commented:
[Test build #37991 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37991/console) for   PR 7554 at commit [`f0c19a0`](https://github.com/apache/spark/commit/f0c19a015778e5086b50398d7e24787f80b68c0d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class DistributedMatrices(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(object):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

mengxr commented:
@dusenberrymw We should split this PR to simplify code review. Try to keep each PR minimal. For example, it would be easier to review if this PR only creates wrappers for existing the Scala APIs. If it is required to add some Scala public API in order to make it work on the Python side, we should have one PR/JIRA for the Scala API, then the Python API. For this PR, I think it should be feasible to have the implementations without `DistribbutedMatrices`.

dusenberrymw commented:
@mengxr Thanks for the thoughts!  I'll trim this PR down to just the Python wrappers, and then open another JIRA up for further discussion on adding a DistributedMatrices class to Scala.

MechCoder commented on docs/mllib-data-types.md:
```diff
@@ -372,6 +372,29 @@ long m = mat.numRows();
+from pyspark.mllib.linalg import DistributedMatrices, Vectors
+
+# Create an RDD of Vectors.
+rows = sc.parallelize([Vectors.dense([1, 2, 3]), Vectors.dense([4, 5, 6]), Vectors.dense([7, 8, 9]), Vectors.dense([10, 11, 12])])
```
Please line break this. Should not be greater than 100.

MechCoder commented on docs/mllib-data-types.md:
```diff
@@ -372,6 +372,29 @@ long m = mat.numRows();
+rows = sc.parallelize([Vectors.dense([1, 2, 3]), Vectors.dense([4, 5, 6]), Vectors.dense([7, 8, 9]), Vectors.dense([10, 11, 12])])
+
+# Create a RowMatrix from an RDD[Vector].
+mat = DistributedMatrices.rowMatrix(rows)
```
+1 to @mengxr 's comment. For now, I think it is sufficient to have just the `RowMatrix`, `IndexedRowMatrix` and `CoordinateMatrix` as separate classes rather than introduce a new class with methods. It improves maintainability.

MechCoder commented on docs/mllib-data-types.md:
```diff
@@ -372,6 +372,29 @@ long m = mat.numRows();
+mat = DistributedMatrices.rowMatrix(rows)
+
+# Get its size.
+m = mat.numRows() # 4
```
nitpick: I think it would be better to just print out the number of rows and columns.

MechCoder commented on docs/mllib-data-types.md:
```diff
@@ -431,7 +454,42 @@ long n = mat.numCols();
+
+An [`IndexedRowMatrix`](api/python/pyspark.mllib.html#pyspark.mllib.linalg.IndexedRowMatrix) can be created from an 
+`RDD` of IndexedRows, where [`IndexedRow`](api/python/pyspark.mllib.html#pyspark.mllib.linalg.IndexedRow) is a 
+wrapper over `(long, Vector)`.  An `IndexedRowMatrix` can be converted to a `RowMatrix` by dropping its row indices.
```
Going by the code samples below, I assume it is not necessary to be long anymore?

MechCoder replied:
Actually it seems Python3 's int is actually long. However doing type(a small integer) gives <type 'int'> and doing type(a large interger) gives <type 'long'> :P

dusenberrymw replied:
Yeah, although it doesn't matter for Python 3, Python 2 still has the distinction, and regardless, from an API/documentation perspective, I think it's a good idea to mimic the requirements of the Scala side, which uses a long in that case.

MechCoder commented on docs/mllib-data-types.md:
```diff
@@ -431,7 +454,42 @@ long n = mat.numCols();
+
+# Create an RDD of indexed rows.
+#   - This can be done explicitly with the IndexedRow class:
+indexedRows = sc.parallelize([IndexedRow(0,Vectors.dense([1, 2, 3])), IndexedRow(1,Vectors.dense([4, 5, 6])), IndexedRow(2,Vectors.dense([7, 8, 9])), IndexedRow(3,Vectors.dense([10, 11, 12]))])
```
please line break and add spaces after commas.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
 
+class DistributedMatrix(object):
+    """Represents a distributively stored matrix backed by one or more RDDs."""
+    def numRows(self):
```
If you would like to represent this as an abstract class, it might be better to use `ABCMeta` and `abstract method` from `abc` module. Check out `pyspark/ml/pipeline.py`

One quick advantage that I can recall from memory is that an error is raised during object creation rather than when the methods are called as it does now. (I think)

dusenberrymw replied:
Well, `Matrix` and `Vector` both follow this same convention, so I think this keeps the codebase consistent.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        # values, which can both be easily serialized.  We will convert back to IndexedRows on the
+        # Scala side.
+        javaIndexedRowMatrix = callMLlibFunc("createIndexedRowMatrix", rows.toDF(),
+                                             long(numRows), int(numCols))
```
I've started using Python3 and long(x) raises a NameError, for the reason mentioned above.

MechCoder replied:
Ah, never mind I see it is renamed in the top of the file.

MechCoder commented on mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala:
```diff
@@ -1105,6 +1108,59 @@ private[python] class PythonMLLibAPI extends Serializable {
+   * Wrapper around DistributedMatrices.rowMatrix factory method.
+   */
+  def createRowMatrix(rows: JavaRDD[Vector], numRows: Long, numCols: Int): RowMatrix = {
+    DistributedMatrices.rowMatrix(rows.rdd, numRows, numCols)
```
I'm wondering out aloud if it would be good to have a Javafriendly constructor in `class RowMatrix` as in many other places.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+
+        :param rows: An RDD of Vectors.
+        """
+        javaRowMatrix = callMLlibFunc("createRowMatrix", rows, long(numRows), int(numCols))
```
Great, it will take minimal amount of effort to just move this over to the class `RowMatrix`

MechCoder replied:
nitpick: You might want to check if rows is a RDD.

dusenberrymw replied:
Yeah should be pretty simple, and yes, checking would be good too.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        self._jirm = jirm
+        self.rows = self._rows()
+
+    def _rows(self):
```
Any reason to make this private? Here and everywhere else. Also I think you should use a property since it is a lass variable on the scala side.

dusenberrymw replied:
Yeah, the `_rows()` method is private since it is just the implementation for the variable `self.rows`, which mimics the `val rows` in Scala.  Pulling the logic out just makes the code cleaner, which is more important for `IndexedRowMatrix` and `CoordinateMatrix` than for `RowMatrix`.

MechCoder replied:
Ah sorry did not see that.

dusenberrymw commented:
@MechCoder Great, thanks for all of the thoughts!  I'll incorporate them as I work on refactoring the logic a bit.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        """
+        Get the rows of the RowMatrix as a RDD of Vectors.
+
+        >>> rows = sc.parallelize([Vectors.dense([1, 2, 3]), Vectors.dense([4, 5, 6]),
```
There are some rules in PEP8 like if we limit the line length to 100 in code, we need to limit the docstring length to 72. I don't understand why, but since they are rules it might be better to follow them :) (http://legacy.python.org/dev/peps/pep-0008/#maximum-line-length)

dusenberrymw replied:
This isn't followed in the rest of the PySpark codebase, but I'd be glad to implement best practices here.

MechCoder replied:
Sure. I realized this from the advice of @mengxr in my previous PR's :)

dusenberrymw replied:
Yeah, thanks for sharing it! :)

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        # DataFrame. This will convert each IndexedRow to a Row containing the 'index' and 'vector'
+        # values, which can both be easily serialized.  We will convert back to IndexedRows on the
+        # Scala side.
+        javaIndexedRowMatrix = callMLlibFunc("createIndexedRowMatrix", rows.toDF(),
```
I do not completely understand why we need to convert this to a dataframe here. Is it just to support non `IndexedRow`s type.

dusenberrymw replied:
The idea is that the `IndexedRow` Python type isn't directly serializable to a format that can be read by the Java side, and vice-versa.  Any of these custom data structures require explicit serialization, which can be found in the `SerDe` Scala class at the bottom of `PythonMLlibAPI.scala`, as well as a bit of representation logic on the Python side.  Prior to `DataFrame`s, this was done for any structure that needed to be passed between the languages, so we already have explicit conversions for `DenseVector`, `DenseMatrix`, `LabeledPoint`, `Rating`, etc.  The `DataFrame` structure lets us pull the contents out of these custom data structures and store them in a a generic `Row`.  The contents are often simpler types, and `DataFrame`s can serialization "primitives" such as `Int`, `Long`, `Boolean`, etc., as well as a few custom types such as `Vector` and `Matrix` types.  So the preferred way of serialization moving forward is to use `DataFrame`s for serialization when possible.  For the `IndexedRow` type, it just contains a `long` and a `Vector`, both of which can be easily serialized with `DataFrame`s, even though the `IndexedRow` itself isn't cross-language serializable.  Converting the `RDD` of `IndexedRow`s in Python to a `DataFrame` creates a `DataFrame` with `Row`s containing just the `long` and `Vector` from the `IndexedRow`.

MechCoder replied:
Thanks for the detailed explanation. But I was also thinking if it would be better to help in adding `IndexedRow` and `MatrixEntity` to SerDe so that it is more useful in the long run?

dusenberrymw replied:
Well, @mengxr mentioned using `DataFrame`s for serialization on the [SPARK-6485 JIRA](https://issues.apache.org/jira/browse/SPARK-6485), so I'm thinking that is the preferred way now.

MechCoder replied:
Great.

MechCoder commented:
@dusenberrymw Thanks for your work! I have made some very minor comments. Also should the structure be `from pyspark.mllib.linalg.distributed import x` ? instead of `from pyspark.mllib.linalg import x`?

dusenberrymw commented:
Thanks again for all of the thoughts!  As for the project structure, I was wondering the same thing.  I know having a package with the same name as the module poses issues in Python, but I suppose we could make a similar change to that of `pyspark.mllib.stat` that appears to use a `_statistics.py` file that is imported entirely when the `stat` package is imported.

@mengxr What are your thoughts on this?

dusenberrymw committed:
Slight reformatting of the documentation.

MechCoder commented on mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala:
```diff
@@ -1105,6 +1108,59 @@ private[python] class PythonMLLibAPI extends Serializable {
+    val indexedRows = rows.map { row =>
+      val index = row.getLong(0)
+      val vector = row.getAs[Vector](1)
+      IndexedRow(index, vector)
```
I am used to writing this in a line like

    rows.map { case Row(ind: Long, vec: Vector) => IndexedRow(ind, vec) }

But that might be personal preference

mengxr commented:
@dusenberrymw For project structure, we can have a separate PR that turns `linalg.py` into a folder `linalg/`. And then you can update this PR (but please address comments first because they become hidden if you move files).

@MechCoder Could you do the refactoring? It is similar to this commit: https://github.com/apache/spark/commit/a3dc6184862345c459d1fba475b1c9210038a913.

dusenberrymw committed:
Updating pattern matching.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38135 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38135/consoleFull) for   PR 7554 at commit [`bb667c4`](https://github.com/apache/spark/commit/bb667c4f5c81cb9349f2706fbd120ad267ee440e).

SparkQA commented:
[Test build #38135 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38135/console) for   PR 7554 at commit [`bb667c4`](https://github.com/apache/spark/commit/bb667c4f5c81cb9349f2706fbd120ad267ee440e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class DistributedMatrices(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(object):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

MechCoder commented:
I have submitted a pull request across your branch, to remove `DistributedMatrices`. Would be great if we could take the discussion there (for now)

dusenberrymw commented:
Thanks, @MechCoder!  Looking forward to discussing more.

dusenberrymw committed:
Pulling the DistributedMatrices Scala class out of this pull request.

dusenberrymw committed:
Pulling the DistributedMatrices Python class out of this pull request.

dusenberrymw committed:
Adding type checks to the constructors.  Although it is slightly verbose, it is better for the user to have a good error message than a cryptic stacktrace.

dusenberrymw committed:
Adding static  methods to each of the distributed matrix classes to consolidate conversion logic.

dusenberrymw committed:
Updating documentation to use class constructors instead of the removed DistributedMatrices factory methods.

MechCoder commented:
Alright, I made some changes to the design.

The main problem was:
RowMatrix initialization needs rows, numRows, numCols for creating a RowMatrix. However while converting from other matrices to a RowMatrix, the java model is used directly.
To prevent this, I have used a `RowMatrixModel`, `RowMatrix` being a wrapper around it, which allows a non-hackish way of initializing.
A minor drawback is that initializing a RowMatrix will give a RowMatrix instance, but converting it from a CoordinateMatrix will give an instance of RowMatrixModel, but we can live with it?

Just in case, you were not notified, this is my Pull Request https://github.com/dusenberrymw/spark/pull/1 . Could you please review it and merge? 

The reason is that the code cutoff is this weekend and it would be great to get this in.

Thanks !

dusenberrymw committed:
Minor updates to documentation.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

dusenberrymw commented:
@MechCoder Thanks for the thoughts!  I too have been working on changes for the past few days since I last pushed updates, and I have pushed those changes now.  Basically, I now have `RowMatrix`, `IndexedRowMatrix`, and `CoordinateMatrix` classes that each take an RDD of `rows` (or `entries` for a `CoordinateMatrix`), optional `numRows`, and optional `numCols`.  For matrix conversions, each has a `_from_java` function that can create a PySpark version of a given Java distributed matrix object, which aids in matrix conversions.  So, if a PySpark `CoordinateMatrix` needs to be converted to a `RowMatrix`, it will first get the resulting `RowMatrix` Java object, then call `RowMatrix._from_java(javaRowMatrix)`, which will then create a PySpark `RowMatrix`.

I think we will want to stay away from having both a `RowMatrix` and a `RowMatrixModel`, especially since a user that converts a `CoordinateMatrix` to a `RowMatrix` will be expecting a `RowMatrix`.  The design I've pushed should keep the API simple and in line with the Scala/Java side.

Interested in any thoughts you may have still!

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        """
+        Create a wrapper over a Java RowMatrix.
+
+        :param rows: An RDD of Vectors.
```
could you document `numRows` and `numCols` here and elsewhere?

dusenberrymw replied:
Yes, thanks.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        >>> int(rm.numRows())
+        4
+
+        >>> rows = sc.parallelize([Vectors.dense([1, 2, 3]), Vectors.dense([4, 5, 6]),
```
nit: I think this can be removed, since `rows` is unchanged? If this be removed, it will also be clearer that the `numRows` and `numCols` depends on the parameters supplied rather than the actual size of RDD's for the same input.

MechCoder replied:
Here and elsewhere

dusenberrymw replied:
Yes, I've kept each of the tests separate up to this point to aid in development, but now is a great time to simplify them.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        self.rows = rows
+
+    @staticmethod
+    def _from_java(javaRowMatrix):
```
For conversion, it seems you are constructing the java_model on the scala side, deconstructing it into rows and constructing the model again in the scala side again. This might be expensive.

I defer to @mengxr for decision regarding this.

dusenberrymw replied:
Yeah I agree, it's not ideal, but it does keep the API clean.  An alternative would be to add an optional Java model parameter to each of the class constructors, and use it if present.  Then we could just construct the Java model on the Scala/Java side, grab it's rows, and pass both of those into the class constructors, bypassing the final step of reconstruction.  This parameter wouldn't be very useful for the end user though.

@mengxr Do you have any thoughts on this?

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        if not isinstance(rows, RDD):
+            raise TypeError("rows should be an RDD object, got %s" % type(rows))
+        first = rows.first()
+        if not isinstance(first, Vector):
```
I suppose it is ok to do

    rows = rows.map(_convert_to_vector)

to support array, list etc types.

dusenberrymw replied:
Yeah it just depends on if we want to enforce the `Vector` requirement.  I'm not opposed to allowing arrays, lists, etc., but if we do, we should document that in the param notes.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        first = rows.first()
+        if not (isinstance(first, IndexedRow) or
+                (isinstance(first, tuple) and len(first) == 2 and
+                 isinstance(first[0], (long, int)) and isinstance(first[1], Vector))):
```
Is it better to convert the map the input to a RDD of IndexedRows? or at least the second element of the tuple to a Vector?

MechCoder replied:
Actually, just doing

    rows.map(lambda row: IndexedRow(*row))

would suffice since there is casting in `IndexedRow`

dusenberrymw replied:
Yeah, I'm not opposed to this, but we should still check to make sure it is 2-element `tuple` so that a useful error message is returned in the event that a completely wrong data structure is passed in.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+            # long to a double for the 3rd tuple value, so map each
+            # tuple to a MatrixEntry, which will do the conversion
+            # on the Python side.
+            entries = entries.map(lambda entry: MatrixEntry(*entry))
```
It seems that there is already typecasting in MatrixEntry as well. You could do away with the checks above.

dusenberrymw replied:
Yeah, same as for the IndexedRows above.  I'm going to create `_convert_to_matrix_entry` function.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+    """
+    Represents a row of an IndexedRowMatrix.
+
+    Just a wrapper over a (long, Vector) tuple.
```
nit: You could move this line over to document `index` and `vector`

dusenberrymw replied:
While fairly verbose, it doesn't hurt.  I'll add it.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+    """
+    Represents an entry of a CoordinateMatrix.
+
+    Just a wrapper over a (long, long, float) tuple.
```
nit: you could move this over to document i, j and value

dusenberrymw replied:
Same as above: while fairly verbose, it doesn't hurt.  I'll add it.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        """
+        if not isinstance(rows, RDD):
+            raise TypeError("rows should be an RDD object, got %s" % type(rows))
+        first = rows.first()
```
You could remove this call here and in CoordinateMatrix by something like this

    def _convert_to_indexed_row(x):
        if isinstance(x, IndexedRow):
            return x
        return IndexedRow(*x)

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+    """
+    Represents a matrix in coordinate format.
+
+    .. note:: Experimental
```
It looks like the convention is to move the Experimental note to the top of the file.

dusenberrymw replied:
Yes, thanks!  (Although I personally like it better at the bottom.)

MechCoder commented:
That's it from my side.

I leave the discussion about conversion to and from different matrices to @mengxr

MechCoder commented:
Oh and would you want to change the project structure yourself or do you want to handle it yourself?

dusenberrymw committed:
Documenting the  and  parameters.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38580 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38580/consoleFull) for   PR 7554 at commit [`ca3302c`](https://github.com/apache/spark/commit/ca3302c9ea5a3b64beaa9f2d9f5b2ce2d36f9942).

SparkQA commented:
[Test build #38580 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38580/console) for   PR 7554 at commit [`ca3302c`](https://github.com/apache/spark/commit/ca3302c9ea5a3b64beaa9f2d9f5b2ce2d36f9942).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(object):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

dusenberrymw committed:
Adding the colons that were accidentally not inserted.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38582 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38582/consoleFull) for   PR 7554 at commit [`059207a`](https://github.com/apache/spark/commit/059207a2a19339633747e3d422807da36a0c5cee).

SparkQA commented:
[Test build #38582 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38582/console) for   PR 7554 at commit [`059207a`](https://github.com/apache/spark/commit/059207a2a19339633747e3d422807da36a0c5cee).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(object):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

dusenberrymw committed:
Simplifying doctests by removing duplicated rows/entries RDDs within the various tests.

dusenberrymw committed:
Moving the Experimental tag to the top of each docstring.

dusenberrymw committed:
Move conversion logic from tuples to IndexedRow or MatrixEntry types from within the IndexedRowMatrix and CoordinateMatrix constructors to separate _convert_to_indexed_row and _convert_to_matrix_entry functions.

dusenberrymw committed:
Adding param documentation to IndexedRow and MatrixEntry.

dusenberrymw committed:
Updating code and documentation to note that a vector-like object (numpy array, list, etc.) can be used in place of explicit Vector object, and adding conversions when necessary to RowMatrix construction.

dusenberrymw commented:
Note: This Python style failure is a build issue due to a dependency update since I started working, rather than an actual issue with the code style.  I'll rebase and push to fix it.

dusenberrymw committed:
Updating doctests to be less verbose by using lists instead of DenseVectors explicitly.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38600 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38600/consoleFull) for   PR 7554 at commit [`a6a7967`](https://github.com/apache/spark/commit/a6a7967025309d87a257655bb77baeca4aa66169).

SparkQA commented:
[Test build #38600 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38600/console) for   PR 7554 at commit [`a6a7967`](https://github.com/apache/spark/commit/a6a7967025309d87a257655bb77baeca4aa66169).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(object):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

dusenberrymw commented:
At this point I still need to update the project structure in a separate PR to convert `linalg.py` to a package.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,392 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        if not isinstance(rows, RDD):
+            raise TypeError("rows should be an RDD of IndexedRows or (long, vector) tuples, "
+                            "got %s" % type(rows))
+        if not isinstance(rows.first(), IndexedRow):
```
Is this check necessary?

It seems that the error handling is done internally in `_convert_to_indexed_row`

dusenberrymw replied:
Yeah, `_convert_to_indexed_row` can handle these cases, but the idea with this check is to avoid the adding an unnecessary conversion job to each element in the case that an RDD of `IndexedRow`s is supplied.

MechCoder replied:
You are right, but there are some examples in spark which are used to handle mixed datatypes, for eg (https://github.com/apache/spark/blob/master/python/pyspark/mllib/feature.py#L237) . This might be highly unlikely, but that's how it is.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,392 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        :param numCols: Number of columns in the matrix.
+        """
+        if not isinstance(rows, RDD):
+            raise TypeError("rows should be an RDD of IndexedRows or (long, vector) tuples, "
```
I think just RDD should be fine in the error check.

MechCoder replied:
I think just RDD should be fine in the error check.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,392 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        Create a wrapper over a Java IndexedRowMatrix.
+
+        :param rows: An RDD of IndexedRows or (long, vector) tuples.
+        :param numRows: Number of rows in the matrix.
```
you could also mention that if it is non-positive, it is set internally (like in the Scala doc)

dusenberrymw replied:
Good idea.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,392 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+
+    Just a wrapper over a (long, vector) tuple.
+
+    :param index: The index for the given row, as a long.
```
sorry for nitpicking, but the user can also give non-long types.

dusenberrymw committed:
Simplifying input conversions in the constructors for each distributed matrix.

dusenberrymw committed:
Updating docstrings.

dusenberrymw committed:
Updating doctests to make documentation cleaner.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38862 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38862/consoleFull) for   PR 7554 at commit [`2cf5d48`](https://github.com/apache/spark/commit/2cf5d4865a37373e12ad876178a54eb0d752ee05).

dusenberrymw committed:
CoordinateMatrix should inherit from DistributedMatrix.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38867 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38867/consoleFull) for   PR 7554 at commit [`a57edbd`](https://github.com/apache/spark/commit/a57edbd0e9c2c238766471faa6e0f05b16b6e70f).

SparkQA commented:
[Test build #38862 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38862/console) for   PR 7554 at commit [`2cf5d48`](https://github.com/apache/spark/commit/2cf5d4865a37373e12ad876178a54eb0d752ee05).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(object):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #38867 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38867/console) for   PR 7554 at commit [`a57edbd`](https://github.com/apache/spark/commit/a57edbd0e9c2c238766471faa6e0f05b16b6e70f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(DistributedMatrix):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

dusenberrymw committed:
Minor documentation cleanup.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38873 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38873/consoleFull) for   PR 7554 at commit [`4ac646b`](https://github.com/apache/spark/commit/4ac646bacc081da6665a7af6a1c6a60de13e590b).

SparkQA commented:
[Test build #38873 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38873/console) for   PR 7554 at commit [`4ac646b`](https://github.com/apache/spark/commit/4ac646bacc081da6665a7af6a1c6a60de13e590b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(DistributedMatrix):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

shaneknapp commented:
sorry, my fault.  i broke the build, and have reverted my change to jenkins.  :)

jenkins, test this please

mengxr commented:
test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38942 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38942/consoleFull) for   PR 7554 at commit [`4ac646b`](https://github.com/apache/spark/commit/4ac646bacc081da6665a7af6a1c6a60de13e590b).

SparkQA commented:
[Test build #38942 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38942/console) for   PR 7554 at commit [`4ac646b`](https://github.com/apache/spark/commit/4ac646bacc081da6665a7af6a1c6a60de13e590b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(DistributedMatrix):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

dusenberrymw committed:
Using properties for better documentation.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38953 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38953/consoleFull) for   PR 7554 at commit [`928c992`](https://github.com/apache/spark/commit/928c99258dd5a6eb461edd76a6ef9ae01f5f70cf).

SparkQA commented:
[Test build #38953 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38953/console) for   PR 7554 at commit [`928c992`](https://github.com/apache/spark/commit/928c99258dd5a6eb461edd76a6ef9ae01f5f70cf).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(DistributedMatrix):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

MechCoder commented on python/pyspark/mllib/linalg.py:
```diff
@@ -1152,9 +1156,416 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+    def __init__(self, rows, numRows=0, numCols=0):
+        """Create a wrapper over a Java RowMatrix."""
+        if not isinstance(rows, RDD):
+            raise TypeError("rows should be an RDD of vectors, got %s" % type(rows))
```
nitpick: it is enough to just say RDD's I think. since the argument need not be vectors (right now).

MechCoder replied:
here and below as well.

dusenberrymw replied:
Yeah the argument doesn't have to be an RDD of actual `Vector` objects, but it should still be an RDD of _vectors_, which could be NumPy arrays, Python lists, `Vector`s, etc. for PySpark. The Spark MLlib Data Types guide makes this distinction for the end-user, so I think it is helpful to use it in the error message as well.

MechCoder commented:
LGTM. cc @mengxr 

We could probably merge this as it is as a start. And leave optimizing the matrix conversions in a future PR.

dusenberrymw commented:
Thanks, @MechCoder!  I say we go ahead and optimize the conversions now though while this is still open. I'm thinking that adding an optional `java_matrix` parameter to the constructors will be the way to go.  Then, if an argument for that is present, we can just store that internally, rather than create a new Java matrix. 

cc @mengxr

mengxr commented:
@dusenberrymw I merged #7761. Please update your PR and put distributed matrices under `pyspark.mllib.linalg.distributed`. Thanks!

dusenberrymw commented:
@mengxr Great! I can finish this tomorrow!

dusenberrymw committed:
Moving the distributed matrices to pyspark.mllib.linalg.distributed.

dusenberrymw committed:
Improving conversion performance.  This adds an optional 'java_matrix' parameter to the constructors, and pulls the conversion logic out into a '_create_from_java' function. Now, if the constructors are given a valid Java distributed matrix object as 'java_matrix', they will store those internally, rather than create a new one on the Scala/Java side.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

dusenberrymw commented:
@mengxr  Okay, I moved the distributed matrices to `pyspark.mllib.linalg.distributed`, and optimized the matrix conversions.  Looking forward to your thoughts!  As soon as this is merged, I can finish #7761.

cc @MechCoder

SparkQA commented:
[Test build #39312 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39312/consoleFull) for   PR 7554 at commit [`687e345`](https://github.com/apache/spark/commit/687e345d7f54b4fe0dbe891e8817ee87421a7c65).

SparkQA commented:
[Test build #39312 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39312/console) for   PR 7554 at commit [`687e345`](https://github.com/apache/spark/commit/687e345d7f54b4fe0dbe891e8817ee87421a7c65).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(DistributedMatrix):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala:
```diff
@@ -1096,6 +1097,56 @@ private[python] class PythonMLLibAPI extends Serializable {
+  def getIndexedRows(indexedRowMatrix: IndexedRowMatrix): DataFrame = {
+    // We use DataFrames for serialization of IndexedRows to Python,
+    // so return a DataFrame.
+    val sqlContext = SQLContext.getOrCreate(SparkContext.getOrCreate())
```
Use `new SQLContext(indexRowMatrix.rows.sparkContext)`. `SQLContext.getOrCreate` doesn't guarantee using the same `SparkContext` as in the input data.

dusenberrymw replied:
Great, I'll change that.

mengxr commented on python/pyspark/mllib/linalg/distributed.py:
```diff
@@ -0,0 +1,513 @@
+#
+
+"""
+MLlib utilities for distributed linear algebra.
```
`MLlib utilities` -> `Package`

mengxr commented on python/pyspark/mllib/linalg/distributed.py:
```diff
@@ -0,0 +1,513 @@
+                    of columns will be determined by the size of
+                    the first row.
+    """
+    def __init__(self, rows, numRows=0, numCols=0, java_matrix=None):
```
We shouldn't expose `java_matrix` in the public API. I suggest the following logic:

1. If `rows` is an Python RDD, convert it to DataFrame and create a Java `RowMatrix` object using `numRows` and `numCols`, only keep the reference to the java object.
2. if `rows` is a Java `RowMatrix`, keep the reference directly. (Do not document it in public API.)

MechCoder replied:
+1. And adding a comment will also help for future development.

dusenberrymw replied:
Yeah, I think that would be really clean, so long as it's documented well internally, as @MechCoder mentioned.

mengxr commented on python/pyspark/mllib/linalg/distributed.py:
```diff
@@ -0,0 +1,513 @@
+            java_matrix = callMLlibFunc("createRowMatrix", rows, long(numRows), int(numCols))
+
+        self._java_matrix_wrapper = JavaModelWrapper(java_matrix)
+        self._rows = rows
```
Continue from previous comment. The issue with keeping both `_java_matrix_wrapper` and `_rows` is that we don't clearly know how they are linked. Keeping only one (the Java object) and use a method to obtain `rows` in Python could simplify the implementation.

dusenberrymw replied:
Yeah if we're going to simplify down to one `rows` parameter that could take the Java object directly, no reason to also compute and store the `rows` anymore.  Previously, the `rows` were directly used to create the Java object, and we already had them, so no reason to get rid of them.

MechCoder replied:
Yes. But it still might be useful to return the rows, entries etc as a `@property`. This could be maybe done by doing `self._rows = rows` initially. not changing `self._rows`, and inside the property returning it as it is if `self._rows` is a RDD and `self._rows.call("rows")` if it is a Java object.

dusenberrymw replied:
Yeah we're definitely keeping a `@property def rows` still, and at a minimum, computing the rows from the Java side when it is called. It would be nice to not waste computation if we already had an RDD to start with on the Python side, but I think the implementation could get messy if we keep checking whether or not rows is an RDD or a JavaObject everywhere.

MechCoder replied:
of course, my comment was specific just to the property.

mengxr commented on python/pyspark/mllib/linalg/distributed.py:
```diff
@@ -0,0 +1,513 @@
+        ...                        [7, 8, 9], [10, 11, 12]])
+
+        >>> rm = RowMatrix(rows)
+        >>> print(rm.numRows())
```
remove `print(..)`

~~~python
>>> rm.numRows()
4
~~~

dusenberrymw replied:
So the issue here is that `numRows()` returns a long, so Python 2 outputs `4L` while Python 3 outputs `4`. We could wrap it in `int()` but I thought that looked more confusing for an end user looking at these tests in the API docs, so I selected `print()`.

mengxr replied:
Okay, then let's keep `print`.

mengxr commented on python/pyspark/mllib/linalg/distributed.py:
```diff
@@ -0,0 +1,513 @@
+                    the first row.
+    """
+    def __init__(self, rows, numRows=0, numCols=0, java_matrix=None):
+        """Create a wrapper over a Java IndexedRowMatrix."""
```
Using `#` for comment? This is not part of the pydoc.

dusenberrymw replied:
Is this referring to the `# We use DataFrames for...` line?  If so, yeah that's only an _internal_ comment to explain why we are using DataFrames for serialization, as it's not immediately clear.  It's not supposed to be part of the public docs.

mengxr replied:
`"""Create a wrapper over a Java IndexedRowMatrix."""` -> `# Create a wrapper over a Java IndexedRowMatrix.`

dusenberrymw replied:
Okay, I can update that, although the PEP 8 & 257 convention is to use docstrings even in the `def __init__(..)` methods.

mengxr replied:
It is okay to use docstring here. But the problem is the content `Create a wrapper over a Java IndexedRowMatrix` is not sufficient for the `__init__` doc. In the generated Python doc, the docstring for the class is used for the constructor rather than this.

mengxr commented on python/pyspark/mllib/linalg/distributed.py:
```diff
@@ -0,0 +1,513 @@
+            java_matrix = callMLlibFunc("createIndexedRowMatrix", rows.toDF(),
+                                        long(numRows), int(numCols))
+
+        self._java_matrix_wrapper = JavaModelWrapper(java_matrix)
```
ditto. Same as in `RowMatrix`. It would be easier to keep only the reference to the Java object.

mengxr commented:
@dusenberrymw Do you have time to update this PR today? Otherwise, we need to postpone it to 1.6.

dusenberrymw committed:
Use 'new SQLContext(matrix.rows.sparkContext)' rather than 'SQLContext.getOrCreate', as the later doesn't guarantee that the SparkContext will be the same as for the matrix.rows data.

dusenberrymw committed:
Updating module docstring.

dusenberrymw commented:
@mengxr Yes, I plan to finish this PR this evening.

dusenberrymw committed:
Updating the matrix conversion logic again to make it even cleaner.  Now, we allow the 'rows' parameter in the constructors to be either an RDD or the Java matrix object. If 'rows' is an RDD, we create a Java matrix object, wrap it, and then store that.  If 'rows' is a Java matrix object of the correct type, we just wrap and store that directly.  This is only for internal usage, and publicly, we still require 'rows' to be an RDD.  We no longer store the 'rows' RDD, and instead just compute it from the Java object when needed.  The point of this is that when we do matrix conversions, we do the conversion on the Scala/Java side, which returns a Java object, so we should use that directly, but exposing 'java_matrix' parameter in the public API is not ideal. This non-public feature of allowing 'rows' to be a Java matrix object is documented in the '__init__' constructor docstrings, which are not part of the generated public API, and doctests are also included.

dusenberrymw committed:
Minor documentation update.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39635 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39635/consoleFull) for   PR 7554 at commit [`bb039cb`](https://github.com/apache/spark/commit/bb039cbb529da4cb4ba21db3fc1cf13ba80a72a5).

dusenberrymw commented:
@mengxr Alright, I've updated this PR.  Now, the `rows` (or `entries`) parameter can (non-publicly) accept a Java matrix object, and we only stored a wrapped Java matrix object for each of the classes.  

Also, since we should still _internally_ document and test this for _internal_ usage, I kept the `__init__` docstrings and expanded them to explain the non-public feature of the `rows` (or `entries`) parameter, as well as to test this feature.  The `__init__` docstrings are not part of the publicly generated docs, so this is a great place to put this.

Thanks, and please let me know if I can update anything else!  Once this is merged, I can finish up #7761.

SparkQA commented:
[Test build #39635 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39635/console) for   PR 7554 at commit [`bb039cb`](https://github.com/apache/spark/commit/bb039cbb529da4cb4ba21db3fc1cf13ba80a72a5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(DistributedMatrix):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

dusenberrymw commented:
Test this please.

dusenberrymw commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39668 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39668/consoleFull) for   PR 7554 at commit [`bb039cb`](https://github.com/apache/spark/commit/bb039cbb529da4cb4ba21db3fc1cf13ba80a72a5).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #198 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/198/consoleFull) for   PR 7554 at commit [`bb039cb`](https://github.com/apache/spark/commit/bb039cbb529da4cb4ba21db3fc1cf13ba80a72a5).

SparkQA commented:
[Test build #198 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/198/console) for   PR 7554 at commit [`bb039cb`](https://github.com/apache/spark/commit/bb039cbb529da4cb4ba21db3fc1cf13ba80a72a5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #39668 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39668/console) for   PR 7554 at commit [`bb039cb`](https://github.com/apache/spark/commit/bb039cbb529da4cb4ba21db3fc1cf13ba80a72a5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(DistributedMatrix):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

dusenberrymw commented:
@mengxr This is completed, but it looks like there is an unrelated Hive test that is failing.  I'll keep trying to test it.

dusenberrymw commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39736 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39736/consoleFull) for   PR 7554 at commit [`bb039cb`](https://github.com/apache/spark/commit/bb039cbb529da4cb4ba21db3fc1cf13ba80a72a5).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #211 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/211/consoleFull) for   PR 7554 at commit [`bb039cb`](https://github.com/apache/spark/commit/bb039cbb529da4cb4ba21db3fc1cf13ba80a72a5).

mengxr commented:
LGTM pending Jenkins.

SparkQA commented:
[Test build #39736 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39736/console) for   PR 7554 at commit [`bb039cb`](https://github.com/apache/spark/commit/bb039cbb529da4cb4ba21db3fc1cf13ba80a72a5).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(DistributedMatrix):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #211 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/211/console) for   PR 7554 at commit [`bb039cb`](https://github.com/apache/spark/commit/bb039cbb529da4cb4ba21db3fc1cf13ba80a72a5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DistributedMatrix(object):`
  * `class RowMatrix(DistributedMatrix):`
  * `class IndexedRow(object):`
  * `class IndexedRowMatrix(DistributedMatrix):`
  * `class MatrixEntry(object):`
  * `class CoordinateMatrix(DistributedMatrix):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

dusenberrymw commented:
@mengxr Alright it looks like one of the Jenkins builds passed all of the tests finally.

mengxr commented:
Great. Merged into master and branch-1.5. Thanks!

dusenberrymw commented:
Thanks, @mengxr!  Yes, I will go finish the block matrix one as well.
